source/retryable-writes/tests/unified/bulkWrite-serverErrors.yml

```
@@ -0,0 +1,96 @@
+description: "retryable-writes bulkWrite serverErrors"
+
+schemaVersion: "1.0"
+
+runOnRequirements:
+  - minServerVersion: "3.6"
+    topologies: [ replicaset ]
+
+createEntities:
+  - client:
+      id: &client0 client0
+      useMultipleMongoses: false
+      observeEvents: [ commandStartedEvent ]
+  - database:
+      id: &database0 database0
+      client: *client0
+      databaseName: &databaseName retryable-writes-tests
+  - collection:
+      id: &collection0 collection0
+      database: *database0
+      collectionName: &collectionName coll
+
+initialData:
+  - collectionName: *collectionName
+    databaseName: *databaseName
+    documents:
+      - { _id: 1, x: 11 }
+      - { _id: 2, x: 22 }
+
+tests:
+  - description: "BulkWrite succeeds after retryable writeConcernError in first batch"
+    runOnRequirements:
+      - minServerVersion: "4.0"
+        topologies: [ replicaset ]
+      - minServerVersion: "4.1.7"
+        topologies: [ sharded-replicaset ]
+    operations:
+      - name: failPoint
+        object: testRunner
+        arguments:
+          client: *client0
+          failPoint:
+            configureFailPoint: failCommand
+            mode: { times: 1 }
+            data:
+              failCommands: [ insert ]
+              errorLabels: [RetryableWriteError] # top-level error labels
+              writeConcernError:
+                code: 91 # ShutdownInProgress
+                errmsg: "Replication is being shut down"
+      - name: bulkWrite
+        object: *collection0
+        arguments:
+          requests:
+            - insertOne:
+                document: { _id: 3, x: 33 }
+            - deleteOne:
+                filter: { _id: 2 }
+        expectResult:
+          deletedCount: 1
+          insertedCount: 1
+          matchedCount: 0
+          modifiedCount: 0
+          upsertedCount: 0
+          insertedIds: { "0": 3 }
```

`insertedIds` in InsertManyResult and BulkWriteResult should be wrapped with `$$unsetOrMatches` since it's optional (see https://github.com/mongodb/specifications/commit/63d3de940b8a8aca43242db6707df84dc472449b for context).

```suggestion
          insertedIds: { $$unsetOrMatches: { 0: 3 } }
```

Also using a numeric literal instead of a string for the key since the map is indexed numerically.

---

Done